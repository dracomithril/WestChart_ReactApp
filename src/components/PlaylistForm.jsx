/**
 * Created by Gryzli on 10.04.2017.
 */
import React, {Component} from "react";
import PropTypes from "prop-types";
import {Button, Checkbox, Form, FormControl, FormGroup, Glyphicon, InputGroup} from "react-bootstrap";
import spotify_utils from "./../spotify_utils";
import utils from "../utils";
const action_types = require('./../reducers/action_types');


export default class PlaylistForm extends Component {
    /*istanbul ignore next*/
    componentWillUnmount() {
        console.log('component PlaylistForm unmounted');
    }

    /*istanbul ignore next*/
    componentDidMount() {
        console.log('component PlaylistForm did mount');
    }

    onStartClick() {
        const {store} = this.context;
        const {selected} = this.props;
        const date = new Date();
        const day1 = date.getDate();
        const str1 = date.toLocaleString('en-US', {month: 'short', day: 'numeric'}).toUpperCase();
        date.setDate(day1 - 4);
        const str2 = date.toLocaleString('en-US', {month: 'short', day: 'numeric'}).toUpperCase();
        let playlist_name = 'Chart_' + str2 + '-' + str1;
        let list = playlist_name.split(' ').join('_');
        store.dispatch({type: action_types.UPDATE_PLAYLIST_NAME, value: list});
        const search = selected.map((elem) => {
            const entry=utils.getArtist_Title(elem.link.title);
            return ({
                artist: entry.artist,
                title: entry.title,
                id: elem.id,
                items: [],
                selected: {}
            });
        });
        store.dispatch({type: action_types.UPDATE_SEARCH, search: search})
    }

    onCreatePlaylist() {
        const {store} = this.context;
        const {search_list, sp_user, sp_playlist_name, isPlaylistPrivate} = store.getState();
        // Create a private playlist
        let update_sp_info = function (spotify_url, sp_name) {
            store.dispatch({type: action_types.UPDATE_PLAYLIST_INFO, value: {url: spotify_url, pl_name: sp_name}});
        };
        const selected = search_list.map((elem) => elem.selected !== undefined ? elem.selected.uri : undefined).filter(elem => elem !== undefined);
        spotify_utils.create_sp_playlist(sp_user, sp_playlist_name, isPlaylistPrivate, selected, update_sp_info);
    }

    render() {
        const {store} = this.context;
        const {sp_playlist_name, isPlaylistPrivate} = store.getState();
        return ( <Form inline>
            <Button onClick={this.onStartClick.bind(this)} id="start_sp_button">Start
            </Button>
            <FormGroup style={{width: 300}} controlId="play_list_name" validationState={((sp_name_length) => {
                return sp_name_length > 8 ? 'success' : sp_name_length > 5 ? 'warning' : 'error';
            })(sp_playlist_name.length)}>
                <InputGroup>
                    <InputGroup.Addon><Glyphicon glyph="music"/></InputGroup.Addon>
                    <FormControl type="text" placeholder="playlist name" value={sp_playlist_name} onChange={(e) => {
                        store.dispatch({
                            type: action_types.UPDATE_PLAYLIST_NAME,
                            value: e.target.value,
                        })
                    }}/>
                    <FormControl.Feedback />
                </InputGroup>
            </FormGroup>
            <FormGroup>
                <Button id="crt_pl_button" onClick={this.onCreatePlaylist.bind(this)}
                        disabled={sp_playlist_name.length < 6}>Create
                    Playlist
                </Button>
                <Checkbox style={{paddingLeft: 5}} id="play_list_is_private" onChange={(e) => {
                    store.dispatch({
                        type: action_types.TOGGLE_IS_PRIVATE,
                        value: e.target.checked
                    })
                }} value="private" checked={isPlaylistPrivate}>{'private ?'}</Checkbox>
            </FormGroup>
        </Form>);
    }
}
PlaylistForm.contextTypes = {
    store: PropTypes.object
};
PlaylistForm.propTypes = {
    selected: PropTypes.array
};