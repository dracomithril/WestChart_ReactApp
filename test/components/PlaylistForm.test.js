/**
 * Created by XKTR67 on 4/19/2017.
 */
import React from 'react';
import PropTypes from 'prop-types';
jest.mock('./../../src/spotify_utils');
import PlaylistForm from '../../src/components/PlaylistForm';
import {shallow,} from 'enzyme';
import {shallowToJson} from 'enzyme-to-json';
import configureMockStore from 'redux-mock-store';
const initial_state = require('../data/initial_state.json');
const data = require('../data/response.json').chart;
const sinon = require('sinon');
const mockStore = configureMockStore([]);

describe('<PlaylistForm/>', () => {
    it('renders without crashing', () => {
        const store = mockStore(initial_state);
        const wrapper = shallow(
            <PlaylistForm />, {
                context: {store},
                childContextTypes: {store: PropTypes.object}
            }
        );
        expect(shallowToJson(wrapper)).toMatchSnapshot();
    });
    it('start_click', () => {
        const state = Object.assign({}, initial_state, {sp_playlist_name: "test_list_zzzz"});
        const store = mockStore(state);
        const wrapper = shallow(
            <PlaylistForm selected={data}/>, {
                context: {store},
                childContextTypes: {store: PropTypes.object}
            }
        );
        expect(shallowToJson(wrapper)).toMatchSnapshot();
        let start_b = wrapper.find('#start_sp_button');
        start_b.simulate('click');
        expect(store.getActions().length).toBe(2)
    });
    it('createPlaylist', () => {

       let sp_use =require('./../../src/spotify_utils');
       sp_use.create_sp_playlist.mockImplementation((a,b,c,f,info)=>{info('aaa','bbb')});
        const state = Object.assign({}, initial_state, {sp_playlist_name: "test_list_zzzz"});
        const store = mockStore(state);
        const wrapper = shallow(
            <PlaylistForm selected={data}/>, {
                context: {store},
                childContextTypes: {store: PropTypes.object}
            }
        );

        let elem = wrapper.find('#play_list_is_private');
        let getElemByIdStub = sinon.stub();
        getElemByIdStub.withArgs('play_list_is_private').returns(elem);
        //document.getElementById = getElemByIdStub;
        expect(shallowToJson(wrapper)).toMatchSnapshot();
        let start_b = wrapper.find('#crt_pl_button');
        start_b.simulate('click');
        expect(store.getActions().length).toBe(1)
    })
});
