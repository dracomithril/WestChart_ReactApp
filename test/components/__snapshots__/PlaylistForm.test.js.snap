// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PlaylistForm/> createPlaylist 1`] = `
<Form
  bsClass="form"
  componentClass="form"
  horizontal={false}
  inline={true}
>
  <Button
    active={false}
    block={false}
    bsClass="btn"
    bsStyle="default"
    disabled={false}
    id="start_sp_button"
    onClick={[Function]}
  >
    Start
  </Button>
  <FormGroup
    bsClass="form-group"
    controlId="play_list_name"
    style={
      Object {
        "width": 300,
      }
    }
    validationState="success"
  >
    <InputGroup
      bsClass="input-group"
    >
      <InputGroupAddon
        bsClass="input-group-addon"
      >
        <Glyphicon
          bsClass="glyphicon"
          glyph="music"
        />
      </InputGroupAddon>
      <FormControl
        bsClass="form-control"
        componentClass="input"
        onChange={[Function]}
        placeholder="playlist name"
        type="text"
        value="test_list_zzzz"
      />
      <FormControlFeedback
        bsClass="form-control-feedback"
        bsRole="feedback"
      />
    </InputGroup>
  </FormGroup>
  <FormGroup
    bsClass="form-group"
  >
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="default"
      disabled={false}
      id="crt_pl_button"
      onClick={[Function]}
    >
      Create Playlist
    </Button>
    <Checkbox
      bsClass="checkbox"
      checked={false}
      disabled={false}
      id="play_list_is_private"
      inline={false}
      onChange={[Function]}
      style={
        Object {
          "paddingLeft": 5,
        }
      }
      value="private"
    >
      private ?
    </Checkbox>
  </FormGroup>
</Form>
`;

exports[`<PlaylistForm/> renders without crashing 1`] = `
<Form
  bsClass="form"
  componentClass="form"
  horizontal={false}
  inline={true}
>
  <Button
    active={false}
    block={false}
    bsClass="btn"
    bsStyle="default"
    disabled={false}
    id="start_sp_button"
    onClick={[Function]}
  >
    Start
  </Button>
  <FormGroup
    bsClass="form-group"
    controlId="play_list_name"
    style={
      Object {
        "width": 300,
      }
    }
    validationState="error"
  >
    <InputGroup
      bsClass="input-group"
    >
      <InputGroupAddon
        bsClass="input-group-addon"
      >
        <Glyphicon
          bsClass="glyphicon"
          glyph="music"
        />
      </InputGroupAddon>
      <FormControl
        bsClass="form-control"
        componentClass="input"
        onChange={[Function]}
        placeholder="playlist name"
        type="text"
        value=""
      />
      <FormControlFeedback
        bsClass="form-control-feedback"
        bsRole="feedback"
      />
    </InputGroup>
  </FormGroup>
  <FormGroup
    bsClass="form-group"
  >
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="default"
      disabled={true}
      id="crt_pl_button"
      onClick={[Function]}
    >
      Create Playlist
    </Button>
    <Checkbox
      bsClass="checkbox"
      checked={false}
      disabled={false}
      id="play_list_is_private"
      inline={false}
      onChange={[Function]}
      style={
        Object {
          "paddingLeft": 5,
        }
      }
      value="private"
    >
      private ?
    </Checkbox>
  </FormGroup>
</Form>
`;

exports[`<PlaylistForm/> start_click 1`] = `
<Form
  bsClass="form"
  componentClass="form"
  horizontal={false}
  inline={true}
>
  <Button
    active={false}
    block={false}
    bsClass="btn"
    bsStyle="default"
    disabled={false}
    id="start_sp_button"
    onClick={[Function]}
  >
    Start
  </Button>
  <FormGroup
    bsClass="form-group"
    controlId="play_list_name"
    style={
      Object {
        "width": 300,
      }
    }
    validationState="success"
  >
    <InputGroup
      bsClass="input-group"
    >
      <InputGroupAddon
        bsClass="input-group-addon"
      >
        <Glyphicon
          bsClass="glyphicon"
          glyph="music"
        />
      </InputGroupAddon>
      <FormControl
        bsClass="form-control"
        componentClass="input"
        onChange={[Function]}
        placeholder="playlist name"
        type="text"
        value="test_list_zzzz"
      />
      <FormControlFeedback
        bsClass="form-control-feedback"
        bsRole="feedback"
      />
    </InputGroup>
  </FormGroup>
  <FormGroup
    bsClass="form-group"
  >
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="default"
      disabled={false}
      id="crt_pl_button"
      onClick={[Function]}
    >
      Create Playlist
    </Button>
    <Checkbox
      bsClass="checkbox"
      checked={false}
      disabled={false}
      id="play_list_is_private"
      inline={false}
      onChange={[Function]}
      style={
        Object {
          "paddingLeft": 5,
        }
      }
      value="private"
    >
      private ?
    </Checkbox>
  </FormGroup>
</Form>
`;
